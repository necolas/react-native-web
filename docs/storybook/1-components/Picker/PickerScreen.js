/* eslint-disable react/jsx-sort-props */

/**
 * @flow
 */

import React from 'react';
import PropOnValueChange from './examples/PropOnValueChange';
import PropSelectedValue from './examples/PropSelectedValue';
import UIExplorer, { AppText, Description, DocItem, Section, storiesOf } from '../../ui-explorer';

const PickerScreen = () => (
  <UIExplorer title="Picker" url="1-components/Picker">
    <Description>
      <AppText>Renders the native &lt;select&gt; component.</AppText>
    </Description>
    <Section title="Props">
      <DocItem
        name="onValueChange"
        typeInfo="?function"
        description="Callback for when an item is selected. This is called with the following parameters: - itemValue: the value prop of the item that was selected - itemPosition: the index of the selected item in this picker."
        example={{
          render: () => <PropOnValueChange />
        }}
      />

      <DocItem
        name="selectedValue"
        typeInfo="?string"
        description="Value matching value of one of the items. Can be a string or an integer."
        example={{
          render: () => <PropSelectedValue />
        }}
      />

      <DocItem name="style" typeInfo="?pickerStyleType" />

      <DocItem
        name="testID"
        typeInfo="?string"
        description="Used to locate this view in end-to-end tests."
      />
    </Section>
  </UIExplorer>
);

storiesOf('Components', module).add('Picker', PickerScreen);
