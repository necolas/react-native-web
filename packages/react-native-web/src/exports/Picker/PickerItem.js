/**
 * Copyright (c) Nicolas Gallagher.
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @noflow
 */

import type { ColorValue } from '../../types';

import createElement from '../createElement';

type Props = {
  color?: ColorValue,
  label: string,
  testID?: string,
  value?: number | string
};

export default function PickerItem(props: Props) {
  const { color, label, testID, value } = props;
  const style = { color };
  return createElement('option', { children: label, style, testID, value });
}
