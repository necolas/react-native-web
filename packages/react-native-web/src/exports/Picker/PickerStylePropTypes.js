/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import ColorPropType from '../ColorPropType';
import ViewStylePropTypes from '../View/ViewStylePropTypes';

const PickerStylePropTypes = {
  ...ViewStylePropTypes,
  color: ColorPropType
};

export default PickerStylePropTypes;
