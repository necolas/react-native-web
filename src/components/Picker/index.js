import Component from './component';
export default Component;
