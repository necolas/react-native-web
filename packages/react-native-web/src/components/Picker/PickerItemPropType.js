/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import React from 'react';
import Picker from './';

const PickerItemPropType = (props: Object, propName: string, componentName: string) => {
  const prop = props[propName];
  let error = null;
  React.Children.forEach(prop, function(child) {
    if (child.type !== Picker.Item) {
      error = new Error('`Picker` children must be of type `Picker.Item`.');
    }
  });
  return error;
};

export default PickerItemPropType;
