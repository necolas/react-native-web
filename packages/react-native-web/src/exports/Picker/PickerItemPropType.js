/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import React from 'react';
import Picker from './';

const PickerItemPropType = (props: Object, propName: string, componentName: string) => {
  const prop = props[propName];
  let error = null;
  React.Children.forEach(prop, function(child) {
    if (child.type !== Picker.Item) {
      error = new Error('`Picker` children must be of type `Picker.Item`.');
    }
  });
  return error;
};

export default PickerItemPropType;
