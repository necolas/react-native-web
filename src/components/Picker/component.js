import React, { Component } from 'react';
import PickerDropDown from './PickerDropDown';
import PickerDialog from './PickerDialog';
import PickerItem from './PickerItem';
import { PICKER_PROPTYPES } from './proptypes';
import {
  MODE_DIALOG,
  MODE_DROPDOWN
} from './constants';

class Picker extends Component {
  static MODE_DIALOG = MODE_DIALOG;

  static MODE_DROPDOWN = MODE_DROPDOWN;

  static defaultProps = {
    mode: MODE_DIALOG
  };

  static propTypes = PICKER_PROPTYPES;

  render() {
    const { mode } = this.props;
    if (mode === MODE_DIALOG) {
      return (<PickerDialog {...this.props} >{this.props.children}</PickerDialog>);
    } else if (mode === MODE_DROPDOWN) {
      return (<PickerDropDown {...this.props} >{this.props.children}</PickerDropDown>);
    }
    return null;
  }
}

/**
 * Individual selectable item in a Picker.
 */
Picker.Item = PickerItem;

export default Picker;

