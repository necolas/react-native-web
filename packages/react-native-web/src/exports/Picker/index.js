/**
 * Copyright (c) 2017-present, Nicolas Gallagher.
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @providesModule Picker
 * @flow
 */

import applyNativeMethods from '../../modules/applyNativeMethods';
import { Component } from 'react';
import ColorPropType from '../ColorPropType';
import createElement from '../createElement';
import PickerItemPropType from './PickerItemPropType';
import StyleSheetPropType from '../../modules/StyleSheetPropType';
import StyleSheet from '../StyleSheet';
import TextPropTypes from '../Text/TextPropTypes';
import ViewPropTypes from '../ViewPropTypes';
import { arrayOf, bool, func, number, oneOfType, string } from 'prop-types';

const pickerStyleType = StyleSheetPropType({
  ...ViewPropTypes.style,
  color: ColorPropType
});

type ItemProps = {
  color?: ColorPropType,
  label: string,
  testID?: string,
  value?: number | string
};

class PickerItem extends Component<ItemProps> {
  static propTypes = {
    color: ColorPropType,
    label: string.isRequired,
    testID: string,
    value: oneOfType([number, string])
  };

  render() {
    const { label, testID, value } = this.props;
    return createElement('option', { label, testID, value });
  }
}

type Props = {
  children?: Array<typeof PickerItem>,
  enabled?: boolean,
  onValueChange?: Function,
  selectedValue?: number | string,
  style?: pickerStyleType,
  testID?: string,
  /* compat */
  itemStyle?: TextPropTypes.style,
  mode?: string,
  prompt?: string
};

class Picker extends Component<Props> {
  static propTypes = {
    children: arrayOf(PickerItemPropType),
    enabled: bool,
    onValueChange: func,
    selectedValue: oneOfType([number, string]),
    style: pickerStyleType,
    testID: string
  };

  static Item = PickerItem;

  render() {
    const {
      children,
      enabled,
      selectedValue,
      style,
      testID,
      /* eslint-disable */
      itemStyle,
      mode,
      prompt
      /* eslint-enable */
    } = this.props;

    return createElement('select', {
      children,
      disabled: enabled === false ? true : undefined,
      onChange: this._handleChange,
      style: [styles.initial, style],
      testID,
      value: selectedValue
    });
  }

  _handleChange = (e: Object) => {
    const { onValueChange } = this.props;
    const { selectedIndex, value } = e.target;
    if (onValueChange) {
      onValueChange(value, selectedIndex);
    }
  };
}

const styles = StyleSheet.create({
  initial: {
    fontFamily: 'inherit',
    fontSize: 'inherit',
    margin: 0
  }
});

export default applyNativeMethods(Picker);
