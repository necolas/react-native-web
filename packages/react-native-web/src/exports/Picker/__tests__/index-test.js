/* eslint-env jasmine, jest */

import React from 'react';
import { shallow } from 'enzyme';
import Picker from '..';

describe('components/Picker', () => {
  describe('prop "children"', () => {
    test('renders items', () => {
      const picker = (
        <Picker>
          <Picker.Item label="label-1" value="value-1" />
          <Picker.Item label="label-2" value="value-2" />
        </Picker>
      );
      const component = shallow(picker);
      expect(component).toMatchSnapshot();
    });
  });

  describe('prop "enabled"', () => {
    test('picker is disabled if false', () => {
      const picker = (
        <Picker enabled={false}>
          <Picker.Item label="label-1" value="value-1" />
          <Picker.Item label="label-2" value="value-2" />
        </Picker>
      );
      const component = shallow(picker);
      expect(component.find('select').props().disabled).toBe(true);
    });
  });

  describe('prop "onValueChange"', () => {
    test('is called with (value, index)', () => {
      const onValueChange = jest.fn();
      const picker = (
        <Picker onValueChange={onValueChange} selectedValue="value-1">
          <Picker.Item label="label-1" value="value-1" />
          <Picker.Item label="label-2" value="value-2" />
        </Picker>
      );
      const component = shallow(picker);
      component.find('select').simulate('change', {
        target: { selectedIndex: '1', value: 'value-2' }
      });
      expect(onValueChange).toHaveBeenCalledWith('value-2', '1');
    });
  });

  describe('prop "selectedValue"', () => {
    test('selects the correct item (string)', () => {
      const picker = (
        <Picker selectedValue="value-2">
          <Picker.Item label="label-1" value="value-1" />
          <Picker.Item label="label-2" value="value-2" />
        </Picker>
      );
      const component = shallow(picker);
      expect(component.find('select').prop('value')).toBe('value-2');
    });

    test('selects the correct item (number)', () => {
      const picker = (
        <Picker selectedValue={22}>
          <Picker.Item label="label-1" value={11} />
          <Picker.Item label="label-2" value={22} />
        </Picker>
      );
      const component = shallow(picker);
      expect(component.find('select').prop('value')).toBe(22);
    });
  });
});
