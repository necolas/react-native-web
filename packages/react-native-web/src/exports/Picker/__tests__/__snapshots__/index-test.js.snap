// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Picker prop "children" renders items 1`] = `
<select
  className="rn-fontFamily-poiln3 rn-fontSize-7cikom rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw"
  onChange={[Function]}
>
  <PickerItem
    label="label-1"
    value="value-1"
  />
  <PickerItem
    label="label-2"
    value="value-2"
  />
</select>
`;
